nixos/modules/misc/nixpkgs.nix

```diff
@@ -416,6 +404,18 @@ in
             ${lib.concatMapStringsSep "\n" (file: "  - ${file}") opt.config.files}
           '';
         }
+        {
+          assertion =
+            (opt.hostPlatform.isDefined -> builtins.isAttrs cfg.buildPlatform -> !(cfg.buildPlatform ? parsed))
```

Comment:
`cfg.buildPlatform` must always be attrs.

Reply:
> > `cfg.buildPlatform` must always be attrs.
>
> it sets `nixpkgs.buildPlatform` to different strings, too.

It could be _defined_ as a string, but previously the `apply` function would always coerce the final value into attrs.

Reply:
I guess what we really need here is a `lib.systems.partiallyElaborate`, so that the `apply` function can be restored without violating the purpose of this PR?

Reply:
> But there is really no need to be able to access `config.nixpkgs.buildPlatform` etc. at all - you can get that information from `pkgs.stdenv.buildPlatform` etc, right?

It seems that up until now, it was intended for users to be able to read those options. But I agree, accessing `pkgs.stdenv` is better (and is what I do in my configs).

I'm trying to think about whether there's a sane way to _enforce_ that though...

Obviously, the option descriptions should be updated with a note explaining they are write-only, and `pkgs.stdenv` should be read instead.

Perhaps an `apply` function could warn or throw when evaluated (similar to mkRemovedOptionModule), but internal usage could bypass the `apply` by merging the option definitions manually?